### PR TITLE
dnsmasq integration

### PR DESCRIPTION
**Pull request by thefosk**
This pull request fixes issue #182, #163 and #127.

The nginx [`resolver`](http://nginx.org/en/docs/http/ngx_http_core_module.html#resolver) property specifies which DNS server nginx will use when resolving addresses.

The problem with the resolver is that by design nginx doesn't follow the system settings including the `/etc/hosts` file. In order to have a proper DNS resolution that respects the system DNS settings a new dependency needs to be introduces: [dnsmasq](http://www.thekelleys.org.uk/dnsmasq/doc.html). 

dnsmasq is a DNS server that runs locally, can listen on a port (I decided to use `8053` for Kong), and can provide DNS resolution reading both `/etc/hosts` file and `resolv.conf`. It is the same dependency that we use for the official Kong Dockerfile (https://github.com/Mashape/docker-kong/blob/master/Dockerfile#L17).

I have been searching for another solution but everybody seems to be using dnsmasq for this use-case.

So this now works, and if in the future a better way to handle this issue will be found, I will be more than happy to remove the dependency to dnsmasq (which works very well by the way). 

I will also update the distributions in a future commit to include dnsmasq as a dependency.

### CODE REVIEW

**Comment by montanaflynn** [comment]
Do you think this may also fix #195?

**Comment by thefosk** [comment]
@montanaflynn not sure, we need to replicate and test that separately after this pull request has been merged into master

**Comment by montanaflynn** [comment]
I've been unable to replicate it, maybe he could build the new version from source and test for us.

**Comment by thefosk** [comment]
The following lines could also be deleted: https://github.com/Mashape/kong/blob/master/kong/dao/cassandra/factory.lua#L27-L38

**Comment by thefosk** [comment]
This pull request is not finalized, I still need to implement some fixes.

**Comment by thibaultCha** [comment]
It looks ok so far, except minor things I'll probably update for the CLI.

- 1) I am very sad, we keep adding more and more dependencies to Kong, and they become more and more expansive (dnsmasq requires a port and runs as yet another process, sigh...)
- 2) I tried playing with [lua-resty-dns](https://github.com/openresty/lua-resty-dns) with no success so far, so I see no other solution.

On a side note:
We should stop relying on the Faker, as it makes the test data very sparse. Some tests end up using the same APIs as others tests and we lose isolation.

We need to use the spec_helper to pass directly through the DAO, and insert only the needed for the current tests at `setup`. And the same APIs need to be dropped at `teardown`.

**Comment by thefosk** [comment]
On a side note I will squash all this commits - basically locally it works, on Travis is failing for some reason.

File: kong/tools/faker.lua [hunk 1]
```diff
@@ -33,6 +33,9 @@ Faker.FIXTURES = {
     { name = "API TESTS 6", public_dns = "cors1.com", target_url = "http://mockbin.com" },
     { name = "API TESTS 7", public_dns = "cors2.com", target_url = "http://mockbin.com" },
 
+    { name = "API TESTS 8 (dns)", public_dns = "dns1.com", target_url = "http://127.0.0.1:7771" },
+    { name = "API TESTS 9 (dns)", public_dns = "dns2.com", target_url = "http://localhost:7771" },
```
Comment (thefosk) [inline kong/tools/faker.lua]: @thibaultCha did you mean like this?

Comment (thibaultCha) [inline kong/tools/faker.lua]: Yeah it's nice it's enough to know it's for the DNS. We'll move them out of the faker and to their own respective tests later.

**Comment by thefosk** [comment]
@thibaultCha if it looks good I can merge

**Comment by thibaultCha** [comment]
I'm having an error when I try to run it:

```
kong|fix/dns ⇒ bin/kong start
[WARN] No configuration at: /etc/kong/kong.yml using default config instead.
[INFO] Using configuration: /usr/local/lib/luarocks/rocks/kong/0.2.1-1/conf/kong.yml
[INFO] Proxy port.........8000
       Admin API port.....8001
       dnsmasq port.......8053
       Database...........cassandra keepalive=60000 hosts=127.0.0.1 port=9042 timeout=1000 keyspace=kong
[INFO] Connecting to the database...
[ERR] /bin/bash: /usr/local/sbin/dnsmasq: No such file or directory
```

**Comment by thefosk** [comment]
@thibaultCha yes, I need to update the distributions but that is in another branch. The new distributions will include dnsmasq.

Also I will need to update the website to list the new requirement, and to document a new configuration property `dnsmasq_port`.

**Comment by thibaultCha** [comment]
Yeah makes sense, it's kinda late here lol, ok nothing to say otherwise, sounds good

**Comment by thefosk** [comment]
Adding some tests to increase the coverage, then I will merge.

**Comment by thefosk** [comment]
Related to https://github.com/Mashape/getkong.org/commit/7c54540c179091c951191abfb1a3656b81570ccb and

**Pull request merged**